src/plugins/intel_gpu/src/graph/impls/onednn/fully_connected_onednn.cpp

```diff
@@ -260,34 +306,84 @@ struct fully_connected_onednn : typed_primitive_onednn_impl<fully_connected> {
+                } else {
+                    // OneDNN does not support scalar zero-point for s4 and u8 type. Need to broadcast it.
+                    auto ifm = arg.get_dependency(1).get_output_layout().get_dim(1);
+                    auto ngroups = arg.get_dependency(decompression_scale_idx).get_output_layout().get_dim(1);
+                    group_size = ifm / ngroups;
+                    attr->set_scales(DNNL_ARG_WEIGHTS, (1 << 1) + (1 << 0), {group_size, 1}, ds_data_type);
+                }
             }
 
             if (prim->decompression_zero_point_scalar.has_value()) {
-                OPENVINO_ASSERT(!prim->decompression_zero_point_scalar.has_value(), "[GPU] OneDNN can't use scalar as a zero point value\n");
+                OPENVINO_ASSERT(!prim->decompression_zero_point_scalar.has_value(), "[GPU] UNIMPLEMENTED: scalar zp support\n");
             } else if (!prim->decompression_zero_point.empty()) {
                 auto decompression_zp_idx = !arg.bias_term() ? 3 : 4;
-                auto data_type = convert_data_type(arg.get_dependency(decompression_zp_idx).get_output_layout().data_type);
-                attr->set_zero_points(DNNL_ARG_WEIGHTS, 1 << 1, dnnl::memory::dims{}, data_type);
+                auto &zp_node = arg.get_dependency(decompression_zp_idx).as<data>();
+                memory::ptr zp_old_mem = zp_node.get_attached_memory_ptr();
+                zp_mem = zp_old_mem;
+
+                if (!is_four_bit_weight) {
+                    // 8-bit quantized weight
+                    auto decompression_zp_idx = !arg.bias_term() ? 3 : 4;
+                    dzp_data_type = convert_data_type(arg.get_dependency(decompression_zp_idx).get_output_layout().data_type);
+                    attr->set_zero_points(DNNL_ARG_WEIGHTS, 1 << 1, dnnl::memory::dims{}, dzp_data_type);
+                } else {
+                    // OneDNN does not support scalar zero-point for s4 and u8 type. Need to broadcast it.
+                    auto decompression_scale_idx = !arg.bias_term() ? 2 : 3;  // it assumes we have decompress_scale
+                    auto &scale_node = arg.get_dependency(decompression_scale_idx).as<data>();
+                    auto broadcasted_layout = scale_node.get_output_layout();
+                    broadcasted_layout.data_type = data_types::u8;
+                    dzp_data_type = convert_data_type(broadcasted_layout.data_type);
+
+                    if (zp_node.get_output_layout().get_linear_size() == 1) {
+                        zp_mem = engine.allocate_memory(broadcasted_layout, false);
+                        auto& stream = engine.get_service_stream();
+                        mem_lock<uint8_t, mem_lock_type::write> zp_new_data(zp_mem, stream);
+                        mem_lock<uint8_t, mem_lock_type::read> zp_old_data(zp_old_mem, stream);
+                        memset(zp_new_data.data(), static_cast<uint8_t>(zp_old_data.data()[0] & 0xf), zp_new_data.size());
```

Comment:
Did you have a chance to evaluate how expensive this broadcasting logic from performance perspective? As we relay on oneDNN kernels cache for dynamic impls, we will create new oneDNN primitive for each new shape. That means that we will do this broadcast operation on the fly for each new shape (also increasing memory consumption by creating new zp broadcasted buffers)
Maybe it would be better to do this broadcasting only once at model loading

Reply:
hmm good point. It is a quick (and easy) implementation. I think proper approach would be to request onednn to support such case without broadcast. Can I follow-up on it after requesting such feature to onednn team?

Reply:
Yeah, I think it's better to have at least this version and then improve broadcasting logic or enable oneDNN's updated implementation